### PR TITLE
issue 1804 - add INDEX_NO_SCHEMA_MATCH

### LINKED ISSUE #1804: Add INDEX_NO_SCHEMA_MATCH module config param
Currently, if no match is found with the fields of a new document, we are still indexing the document and returning it when `FT.SEARCH idx *` is called.
It requires memory in the doc-table and increases query time.

Some users might prefer to not index.

### PR DESCRIPTION
**Pull request by ashtul**
Fix #1804 

This PR changes the default behavior to NOT index documents where no field matches the schema.
It will save space and access time to the doc table.

### CODE REVIEW

**Review by rafie — APPROVED** [review]

**Comment by MeirShpilraien** [comment]
@ashtul @rafie we should discuss if this considered a breaking change

**Review by oshadmi — DISMISSED** [review]
Looks good.

File: src/document.c [hunk 1]
```diff
@@ -116,8 +116,11 @@ static int AddDocumentCtx_SetDocument(RSAddDocumentCtx *aCtx, IndexSpec *sp) {
 
   if (hasTextFields || hasOtherFields) {
     aCtx->stateFlags |= ACTX_F_INDEXABLES;
-  } else {
+  } else if (RSGlobalConfig.indexNoSchemaMatch == 1) {
     aCtx->stateFlags &= ~ACTX_F_INDEXABLES;
+  } else {
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/document.c:121]: The flow has the error in the end which I think is more readable.

[the remark this answers is not in the document]
Comment (oshadmi) [reply src/document.c:121]: Apart from logging it, shouldn't the indexing resume?
Or should it be similar to a scenario such as ```
QUERY_EUNSUPPTYPE
"Tried to index field %s as type not specified in schema"
```

**Comment by ashtul** [comment]
> @ashtul @rafie we should discuss if this considered a breaking change

@MeirShpilraien 
There is a module param that can be set to the old behavior.
Currently, the default is to not index docs w/o any match to the schema.

File: src/document.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (ashtul) [reply src/document.c:121]: @oshadmi 
In this case, there is no indexing to resume. 
Checkout tests where these scenarios are covered.

File: src/document.c [hunk 2]
```diff
@@ -116,8 +116,11 @@ static int AddDocumentCtx_SetDocument(RSAddDocumentCtx *aCtx, IndexSpec *sp) {
...
+    QueryError_SetCode(&aCtx->status, QUERY_ENOMATCH);
+    return -1;
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/document.c:124]: We now handle an error but please note this code is currently dead.

**Review by rafie — DISMISSED** [review]

**Comment by emmanuelkeller** [comment]
- Must be documented
- Is is global or per index ?
- Do we have the use case ?

**Comment by ashtul** [comment]
Not a priority

**Pull request closed without merging**